scrapy_poet/injection.py

```diff
@@ -197,6 +216,37 @@ async def item_factory(page: page_object_cls) -> item_cls:  # type: ignore[valid
 
         return mapping_fn
 
+    @staticmethod
+    def _get_dynamic_deps_factory(
+        dynamic_types: List[type],
+    ) -> Callable[..., DynamicDeps]:
+        """Return a function that creates a :class:`.DynamicDeps` instance from its args.
+
+        It takes instances of types from ``dynamic_types`` as args and returns
+        a :class:`.DynamicDeps` instance where keys are types and values are
+        corresponding args. It has correct type hints so that it can be used as
+        an ``andi`` custom builder.
+        """
+
+        # inspired by dataclasses._create_fn()
+        args = [f"{type_.__name__}_arg: {type_.__name__}" for type_ in dynamic_types]
+        args_str = ", ".join(args)
+        result_args = [
+            f"{type_.__name__}: {type_.__name__}_arg" for type_ in dynamic_types
+        ]
+        result_args_str = ", ".join(result_args)
+        ns = {type_.__name__: type_ for type_ in dynamic_types}
+        create_args = ns.keys()
+        create_args_str = ", ".join(create_args)
+        txt = (
+            f"def __create_fn__({create_args_str}):\n"
+            f" def dynamic_deps_factory({args_str}) -> DynamicDeps:\n"
+            f"  return DynamicDeps({{{result_args_str}}})\n"
+            f" return dynamic_deps_factory"
```

Comment:
What kind of dependency injection is supported for DynamicDeps? E.g. can you have a page object, or an item, or a non-leaf dependency in DynamicDeps? I think we need tests for that.

Reply:
It should support anything that can go to a callback, I'll add some tests.